feat: add upgrade plugin

niladrih opened the pull request:
Changes:
- Renames upgrade crate to openebs-upgrade
- Adds `kubectl openebs upgrade` command
- Adds `kubectl get upgrade-status` command
- Adds Event watcher which watches for initial Upgrade Event

Abhinandan-Purkait commented on plugin/src/cli_utils/upgrade/k8s/mod.rs:
```diff
@@ -0,0 +1,144 @@
+            "'{unsupported_storage_driver}' storage driver for helm is not supported"
+        )),
+    }
+}
```
Add blank line.

Abhinandan-Purkait commented on plugin/src/cli_utils/upgrade/k8s/resources.rs:
```diff
@@ -0,0 +1,316 @@
+        immutable: Some(true),
+        ..Default::default()
+    }
+}
```
Blank line

Abhinandan-Purkait commented on plugin/src/constants.rs:
```diff
@@ -0,0 +1,17 @@
+/// ConfigMap mount path for upgrade.
+pub const UPGRADE_CONFIG_MAP_MOUNT_PATH: &str = "/upgrade-config-map";
+
+pub const PRODUCT_NAME: &str = "openebs";
```
blank line

tiagolobocastro commented on plugin/Cargo.toml:
```diff
@@ -17,23 +17,28 @@ tls = ["openapi/tower-client-tls", "rest-plugin/tls"]
 serde_yaml = "0.9.34"
 snafu = "0.8.5"
-tokio = { version = "1.43.1", features=["full"] }
+tokio = { version = "1.41.0", features=["full"] }
```
why are crates here being downgraded?

tiagolobocastro commented on plugin/src/cli_utils/upgrade/k8s/mod.rs:
```diff
@@ -0,0 +1,144 @@
+    }};
+}
+
+pub fn upgrade_name_concat(release_name: &str, component_name: &str) -> String {
```
add doc comments to pubs

tiagolobocastro commented on plugin/src/cli_utils/upgrade/k8s/mod.rs:
```diff
@@ -0,0 +1,144 @@
+}
+
+
+pub async fn delete_older_upgrade_events(release_name: &str, component_name: &str, kube_event_client: Api<Event>) -> Result<()> {
```
older than what?
add doc comments explaining pls

tiagolobocastro commented on plugin/src/cli_utils/upgrade/mod.rs:
```diff
@@ -0,0 +1,258 @@
+pub mod cli;
+pub mod k8s;
+
+pub struct ImageProperties {
```
What is this for? Add doc comments please

tiagolobocastro commented on plugin/src/cli_utils/upgrade/mod.rs:
```diff
@@ -0,0 +1,258 @@
+    pub async fn new_from_helm_release(ns: &str, release_name: &str, args: &UpgradeArgs) -> Result<Self> {
+        let hostpath_localpv_image_name = format!("{release_name}-localpv-provisioner");
+        let mut openebs_containers: HashSet<&str> = [
+            "api-rest",
```
I guess this would be clearer with comments, what are these for, and why these specifically?

niladrih replied:
Noted, I will add more docs here.

tiagolobocastro commented on plugin/src/cli_utils/upgrade/mod.rs:
```diff
@@ -0,0 +1,258 @@
+                            EnvVar {
+                                // Ref: https://github.com/helm/helm/blob/main/cmd/helm/helm.go#L76
+                                name: "HELM_DRIVER".to_string(),
+                                value: env::var("HELM_DRIVER").ok(),
```
Shouldn't this come from the cli argument?

niladrih replied:
It's being accepted in the same way that helm expects it. Helm introductory docs don't mention this input. I don't expect users to explicitly state this configuration. 

It's expected that the shell is being used for both the plugin and helm.

tiagolobocastro replied:
I mean, you have added the cli args:
```
#[arg(env = "HELM_DRIVER", default_value = "")]
    pub helm_storage_driver: String,
```
But here you're only using the env var, so it's kind pointless to add it to the cli args?

niladrih replied:
Makes sense. It doesn't need to be on the cli args.

niladrih replied:
I have removed the additional env read, and have the CLI arg input everywhere.

tiagolobocastro commented on plugin/src/cli_utils/upgrade/mod.rs:
```diff
@@ -0,0 +1,258 @@
+
+    create_upgrade_resources(args, release_name.as_str(), namespace, k8s_client).await?;
+
+
```
is this formatted? There's a lot of new lines here.

niladrih replied:
It is now..

tiagolobocastro commented on plugin/src/cli_utils/upgrade/mod.rs:
```diff
@@ -0,0 +1,258 @@
+        .map(|_| ())
+        .or_else(|err| match err {
+            kubeError::Api(response) if response.reason.eq("AlreadyExists") => {
+                println!("{kind} '{name}' already exists", kind = resource.kind(), name = resource.name_unchecked());
```
Shouldn't we check the contents match as well?
If the resources are specifically for upgrade, another way could be deleting them before hand, if they exist.

niladrih replied:
The names are unique across versions. So, we should be okay to assume that the resources is the correct one. WDYT?

tiagolobocastro replied:
Then should be ok I guess

tiagolobocastro replied:
Actually it may not be, ex: what if this content is from a previously failed upgrade?

niladrih replied:
There are delete command to delete the resources. And it'll be printed that it already exists, which implies that creation was skipped?

tiagolobocastro commented on plugin/src/constants.rs:
```diff
@@ -0,0 +1,17 @@
+/// The default image tag to the upgrade-job image.
+pub const UPGRADE_JOB_IMAGE_TAG: &str = "develop";
+/// The default image registry for container images.
+pub const DEFAULT_IMAGE_REGISTRY: &str = "docker.io";
```
This is for our upgrade image I'm assuming?

niladrih replied:
Yes!

niladrih committed:
build: add build for openebs-upgrade package and upgrade-job binary

Signed-off-by: Niladri Halder <niladri.halder26@gmail.com>

niladrih committed:
feat: add upgrade plugin

Signed-off-by: Niladri Halder <niladri.halder26@gmail.com>

niladrih commented:
@tiagolobocastro I know quite of bit of doc comments are missing, and they've got a pub on them, so I'll add that in soon, all of the code bits and the build bits are in.

niladrih committed:
build: update release.sh to perform helm dependency update

Signed-off-by: Niladri Halder <niladri.halder26@gmail.com>

abhilashshetty04 commented on plugin/src/main.rs:
```diff
@@ -54,6 +56,14 @@ impl CliArgs {
+                upgrade_args.namespace = ns;
+                upgrade_args.kubeconfig = path
+            }
+            cli_utils::Operations::Get(ref mut get_able) => {
```
Do we need to add `Get` command as a top level verb

Cant it be after `upgrade`

For eg: `kubectl openebs upgrade get status` / ` kubectl openebs upgrade status`

Plugin is developed using `kubectl-openebs resource verb` pattern

tiagolobocastro replied:
> Plugin is developed using kubectl-openebs resource verb pattern

On the mayastor yes, but now on openebs, we have the engine as the first command right?

So it might be ok to keep upgrade as well? 
EDIT: to clarify, I mean using upgrade as top level, as Abhilash showed above

niladrih replied:
I've changed the `kubectl openebs get upgrade-status` to `kubectl openebs upgrade status`. No change to `kubectl openebs upgrade`.

tiagolobocastro commented on plugin/src/cli_utils/upgrade/cli.rs:
```diff
@@ -0,0 +1,70 @@
+
+    /// Helm release name for the openebs helm chart.
+    #[arg(global = true, long, short)]
+    pub release_name: Option<String>,
```
what's the different between namespace and release-name?

niladrih replied:
Kubernetes namespace and helm release name. Example: release 'prod_openebs' in namespace 'storage'

tiagolobocastro commented on plugin/src/cli_utils/upgrade/cli.rs:
```diff
@@ -0,0 +1,70 @@
+    pub release_name: Option<String>,
+
+    /// Specify the container registry for the upgrade-job image.
+    #[arg(global = true, long)]
```
Why are these global?

niladrih replied:
Removed the global attributes

niladrih committed:
style: add doc comments to API components

Signed-off-by: Niladri Halder <niladri.halder26@gmail.com>

niladrih committed:
feat: change command tree for upgrade
This changes the `kubectl openebs get upgrade-status` to `kubectl openebs upgrade status`.

Signed-off-by: Niladri Halder <niladri.halder26@gmail.com>
